contribs/analysis/src/main/java/org/matsim/contrib/analysis/time/TimeBinMap.java

```diff
@@ -14,111 +15,121 @@
-        }
-
-        public void setValue(T value) {
-            this.value = value;
-        }
-
-        public boolean hasValue() {
-            return this.value != null;
-        }
-    }
-}
+	private final Map<Integer, TimeBin<T>> bins = new HashMap<>();
+	private final double binSize;
+	private final double startTime;
+	private double endTimeOfLastBucket;
+
+	/**
+	 * Creates new instance of TimeBinMap
+	 *
+	 * @param timeBinSize size of one time bin
+	 */
+	public TimeBinMap(final double timeBinSize) {
+		this(timeBinSize, 0);
+	}
+
+	/**
+	 * Creates a new instance of TimeBinMap
+	 *
+	 * @param timeBinSize         size of one time bin
+	 * @param startTimeOfFirstBin start time of first time bin. Default is 0.
+	 */
+	public TimeBinMap(final double timeBinSize, final double startTimeOfFirstBin) {
+		this.binSize = timeBinSize;
+		this.startTime = startTimeOfFirstBin;
+	}
+
+	/**
+	 * Get a time bin for the given time. Returns an existing time bin or creates a new one if no bin is present yet.
+	 *
+	 * @param forTime time for which a bin is requested.
+	 * @return Time bin which contains the requested time
+	 */
+	public TimeBin<T> getTimeBin(double forTime) {
+
+		if (forTime < startTime)
+			throw new IllegalArgumentException("only times greater than startTime of: " + startTime + " allowed");
+
+		int binIndex = (int) ((forTime - startTime) / binSize);
+		if (!bins.containsKey(binIndex)) {
+			bins.put(binIndex, new TimeBin<>(startTime + binIndex * binSize));
+			adjustEndTime(binIndex);
+		}
+		return bins.get(binIndex);
+	}
+
+	/**
+	 * Retrieve the end time of the last time bin.
+	 *
+	 * @return end time of the last bin and of the whole time series stored by this map
+	 */
+	public double getEndTimeOfLastBin() {
+		return endTimeOfLastBucket;
+	}
+
+	public double getStartTime() {
+		return startTime;
+	}
+
+	public double getBinSize() {
+		return binSize;
+	}
+
+	/**
+	 * Retrieve all time bins. Note: There is no guarantee that time bins are consecutive. Since only requested bins are
+	 * created, it is possible that there are 'gaps' in the time series.
+	 *
+	 * @return all time bins
+	 */
+	public Collection<TimeBin<T>> getTimeBins() {
+		return bins.values();
+	}
+
+	/**
+	 * Delete all values
+	 */
+	public void clear() {
+		bins.clear();
+	}
+
+	private void adjustEndTime(int newBinIndex) {
+
+		if (startTime + binSize + newBinIndex * binSize > endTimeOfLastBucket)
+			endTimeOfLastBucket = startTime + binSize + newBinIndex * binSize;
+	}
+
+	public static class TimeBin<T> {
+
+		private final double startTime;
+		private T value;
+
+		private TimeBin(double startTime) {
+			this.startTime = startTime;
+		}
+
+		public double getStartTime() {
+			return startTime;
+		}
+
+		public T getValue() {
+			return value;
+		}
+
+		public void setValue(T value) {
+			this.value = value;
+		}
+
+		public T getValue(Supplier<T> computeIfAbsent) {
+
+			if (!hasValue()) {
+				setValue(computeIfAbsent.get());
```

Comment:
The name of the method indicates no side effect behaviour (like `Map.getOrDefault()`), whereas there is potentially one. I suggest changing the name to `computeIfAbsent()` (following the Java Collections nomenclature).

Reply:
Thanks for the suggestion. I will do so!